prusti-specs/src/extern_spec_rewriter/common.rs

```diff
@@ -142,11 +143,40 @@ impl<'a> AssociatedTypeRewriter<'a> {
     }
 
     pub fn rewrite_method_sig(&mut self, signature: &mut syn::Signature) {
-        syn::visit_mut::visit_signature_mut(self, signature);
+        self.visit_signature_mut(signature);
+    }
+
+    pub fn rewrite_attribute(&mut self, attr: &mut syn::Attribute) {
+        self.visit_attribute_mut(attr);
+    }
+
+    fn rewrite_tokens(&self, self_replacement_as_str: &str, tokens: TokenStream) -> TokenStream {
+        let it = tokens.into_iter();
+        it.map(|tt| {
+            match tt {
+                TokenTree::Ident(ident) if ident == "Self" => {
+                    TokenTree::Ident(Ident::new(self_replacement_as_str, ident.span()))
+                }
+                TokenTree::Group(group) => {
+                    TokenTree::Group(Group::new(group.delimiter(), self.rewrite_tokens(self_replacement_as_str, group.stream())))
+                },
+                other => other,
+            }
+        }).collect()
     }
 }
 
 impl<'a> syn::visit_mut::VisitMut for AssociatedTypeRewriter<'a> {
+    fn visit_attribute_mut(&mut self, attr: &mut syn::Attribute) {
+        let self_replacement = self.self_type.to_token_stream().to_string();
+
+        // Note: Attribute tokens are not visited by a syn visitor,
+        // so we visit the attribute tokens "manually"
+        attr.tokens = self.rewrite_tokens(self_replacement.as_str(), attr.tokens.clone());
```

Comment:
I think in order to do that we need to process the nested specs in an extern spec with `generate_spec_and_assertions` like we do for trait spec refinements.
I think this is generally a good idea.
Can we do this after we merged abstract predicates #938? I moved this logic to a new place and would like to avoid nasty merges. So my suggestion:
- Leave this as it is rn
- Finish abstract predicates
- New PR where we eagerly process nested specs in extern specs